typing/typeopt.ml

```diff
@@ -168,9 +168,41 @@ let value_kind env ty =
       Pboxedintval Pint64
   | Tconstr(p, _, _) when Path.same p Predef.path_nativeint ->
       Pboxedintval Pnativeint
+  | Tconstr(p, _, _) ->
+      if Numbers.Int.Set.mem ty.id visited then
+        Pgenval
+      else begin
+        match Env.find_type_descrs p env with
+        | exception Not_found -> Pgenval
+        | [], _ -> Pgenval
+        | [ { cstr_tag = Cstr_block { tag }; cstr_args } ], _ ->
+          let visited = Numbers.Int.Set.add ty.id visited in
+          let fields = List.map (value_kind' env ~visited) cstr_args in
+          Pblock { tag; fields }
+        | constructors, _ ->
+          let is_constant = function
+            | { cstr_tag = Cstr_constant _ } -> true
+            | _ -> false
+          in
+          if List.for_all is_constant constructors then
+            Pintval
+          else
+            Pgenval
+      end
+  | Ttuple fields ->
+      if Numbers.Int.Set.mem ty.id visited then
+        Pgenval
+      else begin
+        let visited = Numbers.Int.Set.add ty.id visited in
+        let fields = List.map (value_kind' env ~visited) fields in
+        Pblock { tag = 0; fields }
+      end
```

Comment:
We need to find someone who understands the type checker to review this part.  Perhaps ask @lpw25 .